pkg/config/remote/service/subscriber.go

```diff
@@ -0,0 +1,125 @@
+	"crypto/tls"
+	"fmt"
+	"io"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/metadata"
+
+	"github.com/DataDog/datadog-agent/pkg/api/security"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/config/remote/service/tuf"
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// SubscriberCallback defines the function called when a new configuration was fetched
+type SubscriberCallback func(config *pbgo.ConfigResponse) error
+
+// Subscriber describes a product's configuration subscriber
+type Subscriber struct {
+	product     pbgo.Product
+	refreshRate time.Duration
+	lastUpdate  time.Time
+	lastVersion uint64
+	callback    SubscriberCallback
+}
+
+// NewSubscriber returns a new subscriber with the specified refresh rate and a callback
+func NewSubscriber(product pbgo.Product, refreshRate time.Duration, callback SubscriberCallback) *Subscriber {
+	return &Subscriber{
+		product:     product,
+		refreshRate: refreshRate,
+		callback:    callback,
+	}
+}
+
+// NewGRPCSubscriber returns a new gRPC stream based subscriber.
+func NewGRPCSubscriber(product pbgo.Product, callback SubscriberCallback) (context.CancelFunc, error) {
+	creds := credentials.NewTLS(&tls.Config{
+		InsecureSkipVerify: true,
+	})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	conn, err := grpc.DialContext(
+		ctx,
+		fmt.Sprintf(":%v", config.Datadog.GetInt("cmd_port")),
+		grpc.WithTransportCredentials(creds),
+	)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	agentClient := pbgo.NewAgentSecureClient(conn)
+
+	token, err := security.FetchAuthToken()
+	if err != nil {
+		cancel()
+		err = fmt.Errorf("unable to fetch authentication token: %w", err)
+		log.Infof("unable to establish stream, will possibly retry: %s", err)
+		return nil, err
+	}
+
+	currentConfigSnapshotVersion := uint64(0)
+	client := tuf.NewDirectorPartialClient()
+
+	go func() {
+		log.Debug("Waiting for configuration from remote config management")
+
+		streamCtx, streamCancel := context.WithCancel(
+			metadata.NewOutgoingContext(ctx, metadata.MD{
+				"authorization": []string{fmt.Sprintf("Bearer %s", token)},
+			}),
+		)
+		defer streamCancel()
+
+		log.Debug("Processing response from remote config management")
+
+		for {
+			request := pbgo.SubscribeConfigRequest{
+				CurrentConfigSnapshotVersion: currentConfigSnapshotVersion,
+				Product:                      product,
+			}
+			stream, err := agentClient.GetConfigUpdates(streamCtx, &request)
+			if err != nil {
+				log.Error("Failed to request configuration, retrying in 3 seconds...")
+				time.Sleep(3 * time.Second)
+				continue
+			}
+
+			for {
+				// Get new event from stream
+				configResponse, err := stream.Recv()
+				if err == io.EOF {
+					continue
+				} else if err != nil {
+					log.Warnf("Stopped listening for configuration from remote config management: %s", err)
```

Comment:
I was running this code with a wrong configuration which was producing a big amount of logs. We might want to add a sleep or ticker here that only checks the stream every X seconds.

Reply:
I am not sure why it would not go through the `time.Sleep(3 * time.Second)` line 97. It successfully connects for fails while reading messages ? Anyway, I agree it makes sense to sleep a bit to avoid having too many logs.

Reply:
Indeed, it's weird. I am getting a `Stopped listening for configuration from remote config management: rpc error: code = Unknown desc = remote config service not initialized` in line 103 that somehow falls through

Reply:
That must happen when remote config is not enabled in the agent. For some reason, connection works but updates can be read obviously.

Reply:
I reproduced locally and verified the issue is fixed